File: algo/dp/lc879.java
package algo.dp;

/**
 * 1、01背包问题的变形
 * 2、状态 dp[i][j][k] 只用前i个任务，不超过j个人，获得的利润不小于k的方案数
 */
public class lc879 {
    public int profitableSchemes(int n, int minProfit, int[] group, int[] profit) {
        int len = group.length, MOD = (int)1e9 + 7;
        int[][][] dp = new int[len + 1][n + 1][minProfit + 1];
        dp[0][0][0] = 1;
        for (int i = 1; i <= len; i++) {
            int members = group[i - 1], earn = profit[i - 1];
            for (int j = 0; j <= n; j++) {
                for (int k = 0; k <= minProfit; k++) {
                    if (j < members) {
                        dp[i][j][k] = dp[i - 1][j][k];
                    } else {
                        dp[i][j][k] = (dp[i - 1][j][k] + dp[i - 1][j - members][Math.max(0, k - earn)]) % MOD;
                    }
                }
            }
        }
        int sum = 0;
        for (int j = 0; j <= n; j++) {
            sum = (sum + dp[len][j][minProfit]) % MOD;
        }
        return sum;
    }
    public int profitableSchemes_2(int n, int minProfit, int[] group, int[] profit) {
        int len = group.length, MOD = (int)1e9 + 7;
        int[][] dp = new int[n + 1][minProfit + 1];
        dp[0][0] = 1;
        for (int i = 1; i <= len; i++) {
            int members = group[i - 1], earn = profit[i - 1];
            for (int j = n; j >= 0; j--) {
                for (int k = 0; k <= minProfit; k++) {
                    if (j >= members) {
                        dp[j][k] = (dp[j][k] + dp[j - members][Math.max(0, k - earn)]) % MOD;
                    }
                }
            }
        }
        int sum = 0;
        for (int j = 0; j <= n; j++) {
            sum = (sum + dp[j][minProfit]) % MOD;
        }
        return sum;
    }
}

File: hard/Lc879/Lc879.py
# 集团里有 n 名员工，他们可以完成各种各样的工作创造利润。

# 第 i 种工作会产生 profit[i] 的利润，它要求 group[i] 名成员共同参与。如果成员参与了其中一项工作，就不能参与另一项工作。

# 工作的任何至少产生 minProfit 利润的子集称为 盈利计划 。并且工作的成员总数最多为 n 。

# 有多少种计划可以选择？因为答案很大，所以 返回结果模 10^9 + 7 的值。
class Solution:
    def profitableSchemes(self, n: int, minProfit: int, g: List[int], p: List[int]) -> int:
        MOD = 10**9 + 7
        ans, m = 0, len(g)
        dp = [[0] * (minProfit + 1) for _ in range(n + 1)]
        dp[0][0] = 1
        for i in range(1,m+1):
            for j in range(n, -1, -1):
                for k in range(minProfit, -1, -1):
                    if 0 <= j-g[i-1]:
                        dp[j][k] = (dp[j][k] + dp[j-g[i-1]][max(0,k-p[i-1])])%MOD
        for j in range(n+1):
            ans = (ans + dp[j][minProfit])%MOD
        return ans%MOD